File: Extendtablayout/src/main/java/com/summer/extendtablayout/ExtendTabLayout.java
package com.summer.extendtablayout;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.StateListDrawable;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.google.android.material.tabs.TabLayout;

/**
 * 一个继承于{@link TabLayout}
 * 通过customView的方式改变tabLayout的indicator，让indicator的宽度等于文字的宽度
 * 且 可以通过属性设置indicator距离tabLayout的左侧右侧距离
 */
public class ExtendTabLayout extends TabLayout implements TabLayout.BaseOnTabSelectedListener {

    private final int indicatorHeight;
    @ColorInt
    @ColorRes
    private int indicatorSelectColor;

    private final int indicatorWidth;//指示器宽度

    private @DrawableRes
    int indicatorDrawable;

    private float tabTextSize, tabSelectTextSize, defaultTextSize;

    private final int indicatorMarginStart;

    private final int indicatorMarginEnd;

    private final int indicatorOverStart, indicatorOverEnd;

    public ExtendTabLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, com.google.android.material.R.attr.tabStyle);
    }

    public ExtendTabLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.ExtendTabLayout);
        indicatorHeight = array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabIndicatorHeight, 0);
        indicatorSelectColor = array.getColor(R.styleable.ExtendTabLayout_extendTabIndicatorSelectColor, 0);
        indicatorMarginStart = array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabIndicatorMarginStart, 0);
        indicatorMarginEnd = array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabIndicatorMarginEnd, 0);
        tabTextSize = (float) array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabTextSize, 0);
        indicatorDrawable = array.getResourceId(R.styleable.ExtendTabLayout_extendTabIndicatorDrawable, 0);
        indicatorWidth = array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabIndicatorWidth, 0);
        indicatorOverStart = array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabIndicatorOverStart, 0);
        indicatorOverEnd = array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabIndicatorOverEnd, 0);
        tabSelectTextSize = (float) array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabSelectTextSize, 0);
        array.recycle();

        addOnTabSelectedListener(this);
    }


    @Override
    public void addTab(@NonNull Tab tab, int position, boolean setSelected) {
        tab.setCustomView(createTabCustomView(tab.getText(), tab.view));
        super.addTab(tab, position, setSelected);
    }

    /**
     * 设置tab
     */
    private View createTabCustomView(CharSequence text, View tabView) {
        View view = LayoutInflater.from(getContext()).inflate(R.layout.layout_extentd_tab, (LinearLayout) tabView, false);
        TextView textView = view.findViewById(R.id.layout_extend_tab_tv_content);
        textView.setText(text);
        textView.setTextColor(getTabTextColors());
        if (tabTextSize != 0f) {
            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, tabTextSize);
        } else {
            defaultTextSize = textView.getTextSize();
        }
        textView.setPadding(indicatorOverStart, 0, indicatorOverEnd, 0);
        View indicator = view.findViewById(R.id.layout_extend_tab_indicator);
        ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) indicator.getLayoutParams();
        params.height = indicatorHeight;
        params.leftMargin = indicatorMarginStart;
        params.rightMargin = indicatorMarginEnd;
        params.width = indicatorWidth;
        indicator.setLayoutParams(params);

        indicator.setBackground(createStateListDrawable());

        return view;
    }

    /**
     * 创建StateListDrawable
     *
     * @return
     */
    private StateListDrawable createStateListDrawable() {
        StateListDrawable stateListDrawable = new StateListDrawable();
        Drawable drawable = null;
        if (indicatorDrawable == 0) {
            drawable = new ColorDrawable(indicatorSelectColor);
        } else {
            drawable = getResources().getDrawable(indicatorDrawable);
        }
        stateListDrawable.addState(new int[]{android.R.attr.state_selected}, drawable);
        stateListDrawable.addState(new int[]{}, new ColorDrawable());
        return stateListDrawable;
    }

    @Override
    public void onTabSelected(Tab tab) {
        if (((int) tabSelectTextSize) == 0) return;
        changeTabTextSize(tab, tabSelectTextSize);
    }

    @Override
    public void onTabUnselected(Tab tab) {

    }

    @Override
    public void onTabReselected(Tab tab) {
        if (((int) tabSelectTextSize) == 0) return;
        changeTabTextSize(tab, ((int) tabTextSize) == 0 ? defaultTextSize : tabTextSize);
    }

    /**
     * 改变tab的字体大小
     *
     * @param tab
     * @param textSize
     */
    private void changeTabTextSize(Tab tab, float textSize) {
        View view = tab.getCustomView();
        if (view == null) return;
        TextView tv = view.findViewById(R.id.layout_extend_tab_tv_content);
        if (tv != null) {
            tv.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize);
        }
    }
}

File: Extendtablayout/src/main/java/com/summer/extendtablayout/ExtendTabLayout.kt
package com.summer.extendtablayout

import android.content.Context
import kotlin.jvm.JvmOverloads
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import android.view.LayoutInflater
import android.widget.LinearLayout
import android.widget.TextView
import android.util.TypedValue
import androidx.constraintlayout.widget.ConstraintLayout
import android.graphics.drawable.Drawable
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.StateListDrawable
import android.util.AttributeSet
import android.util.Log
import android.util.SparseArray
import android.view.View
import android.widget.FrameLayout
import com.google.android.material.tabs.Util
import com.summer.extendtablayout.util.TabIndicatorInterpolatorFactory
import org.w3c.dom.Text
import java.lang.reflect.Field

/**
 * 一个继承于[TabLayout]
 * 通过customView的方式改变tabLayout的indicator，让indicator的宽度等于文字的宽度
 * 且 可以通过属性设置indicator距离tabLayout的左侧右侧距离
 *
 * 2021年11月15日更新  为了切换时Indicator动画能够正常执行，修改原生的Indicator,而不采用自定义CustomView的方式
 */
const val DEFAULT_VALUE = 0
class ExtendTabLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyleAttr: Int = com.google.android.material.R.attr.tabStyle
) : TabLayout(context, attrs, defStyleAttr), OnTabSelectedListener {

    private val indicatorWidth: Int //指示器宽度

    private val tabSelectTextSize: Int
    private val indicatorMarginStart: Int
    private val indicatorMarginEnd: Int
    private val indicatorOverStart: Int
    private val indicatorOverEnd: Int

    private var tabIndicatorAnimationMode: Int? = null


    init {
        val array = context.obtainStyledAttributes(attrs, R.styleable.ExtendTabLayout)
        indicatorMarginStart = array.getDimensionPixelSize(
            R.styleable.ExtendTabLayout_extendTabIndicatorMarginStart,
            DEFAULT_VALUE
        )
        indicatorMarginEnd = array.getDimensionPixelSize(
            R.styleable.ExtendTabLayout_extendTabIndicatorMarginEnd,
            DEFAULT_VALUE
        )
        indicatorWidth =
            array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabIndicatorWidth, DEFAULT_VALUE)
        indicatorOverStart = array.getDimensionPixelSize(
            R.styleable.ExtendTabLayout_extendTabIndicatorOverStart,
            DEFAULT_VALUE
        )
        indicatorOverEnd = array.getDimensionPixelSize(
            R.styleable.ExtendTabLayout_extendTabIndicatorOverEnd,
            DEFAULT_VALUE
        )
        tabSelectTextSize =
            array.getDimensionPixelSize(R.styleable.ExtendTabLayout_extendTabSelectTextSize, DEFAULT_VALUE)
        array.recycle()
        tabIndicatorAnimationMode?.apply { setTabIndicatorAnimationMode(this) }
        if (tabSelectTextSize != DEFAULT_VALUE) {
            addOnTabSelectedListener(this)
        }
    }

    override fun addTab(tab: Tab, position: Int, setSelected: Boolean) {
        if (tabSelectTextSize != DEFAULT_VALUE) {
            tab.customView = Util.createCustomView(context, this)
        }
        super.addTab(tab, position, setSelected)
    }



    override fun setTabIndicatorAnimationMode(tabIndicatorAnimationMode: Int) {
        this.tabIndicatorAnimationMode = tabIndicatorAnimationMode
        if (indicatorMarginStart != DEFAULT_VALUE || indicatorMarginEnd != DEFAULT_VALUE || indicatorOverStart != DEFAULT_VALUE || indicatorOverEnd != DEFAULT_VALUE) {
            if (!TabIndicatorInterpolatorFactory.createTabIndicatorInterpolator(
                    this,
                    indicatorMarginStart,
                    indicatorMarginEnd,
                    indicatorOverStart,
                    indicatorOverEnd,
                    indicatorWidth
                )
            ) {
                super.setTabIndicatorAnimationMode(tabIndicatorAnimationMode)
            }
            return
        }
        super.setTabIndicatorAnimationMode(tabIndicatorAnimationMode)
    }

    override fun onTabSelected(tab: Tab) {
        if (tabSelectTextSize == 0) return
        changeTextSize(tab, tabSelectTextSize.toFloat())

    }

    override fun onTabUnselected(tab: Tab) {
        changeTextSize(tab, Util.getTabTextSize(this))
    }
    override fun onTabReselected(tab: Tab) {
        if (tabSelectTextSize == 0) return
        changeTextSize(tab, tabSelectTextSize.toFloat())
    }

    /**
     * 通过反射改变文字大小
     */
    private fun changeTextSize(tab: Tab, tabSelectTextSize: Float) {
        tab.customView?.apply {
            val textView =findViewById<TextView>(android.R.id.text1)
            textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, tabSelectTextSize)
        }
    }


}